cli/commands/scaffold/action.go

```diff
@@ -0,0 +1,351 @@
+  # Optional input variables
+  # Uncomment the ones you wish to set
+  # --------------------------------------------------------------------------------------------------------------------
+  {{ range .optionalVariables }}
+  # Description: {{ .Description }}
+  # Type: {{ .Type }}
+  # {{ .Name }} = {{ .DefaultValue }}
+  {{ end }}
+}
+`
+)
+
+var moduleUrlRegex = regexp.MustCompile(moduleUrlPattern)
+
+func Run(opts *options.TerragruntOptions) error {
+	// download remote repo to local
+	var moduleUrl = ""
+	var templateUrl = ""
+	// clean all temp dirs
+	var dirsToClean []string
+	defer func() {
+		for _, dir := range dirsToClean {
+			if err := os.RemoveAll(dir); err != nil {
+				opts.Logger.Warnf("Failed to clean up dir %s: %v", dir, err)
+			}
+		}
+	}()
+
+	// scaffold only in empty directories
+	if empty, err := util.IsDirectoryEmpty(opts.WorkingDir); !empty || err != nil {
+		if err != nil {
+			return err
+		}
+		return WorkingDirectoryNotEmptyError{
+			dir: opts.WorkingDir,
+		}
+	}
+
+	if len(opts.TerraformCliArgs) >= moduleCount {
+		moduleUrl = opts.TerraformCliArgs[1]
+	}
+
+	if len(opts.TerraformCliArgs) >= moduleAndTemplateCount {
+		templateUrl = opts.TerraformCliArgs[2]
+	}
+
+	tempDir, err := os.MkdirTemp("", "scaffold")
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	dirsToClean = append(dirsToClean, tempDir)
+
+	// prepare variables
+	vars, err := variables.ParseVars(opts.ScaffoldVars, opts.ScaffoldVarFiles)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// parse module url
+	parsedModuleUrl, err := terraform.ToSourceUrl(moduleUrl, tempDir)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	moduleUrl = parsedModuleUrl.String()
+
+	// rewrite module url, if required
+	parsedModuleUrl, err = rewriteModuleUrl(opts, vars, moduleUrl)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// add ref to module url, if required
+	parsedModuleUrl, err = addRefToModuleUrl(opts, parsedModuleUrl, vars)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// regenerate module url with all changes
+	moduleUrl = parsedModuleUrl.String()
+
+	// identify template url
+	templateDir := ""
+	if templateUrl != "" {
+		// process template url if was passed
+		parsedTemplateUrl, err := terraform.ToSourceUrl(templateUrl, tempDir)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		parsedTemplateUrl, err = rewriteTemplateUrl(opts, parsedTemplateUrl)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		// regenerate template url with all changes
+		templateUrl = parsedTemplateUrl.String()
+
+		templateDir, err = os.MkdirTemp("", "template")
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		dirsToClean = append(dirsToClean, templateDir)
+
+		// downloading template
+		opts.Logger.Infof("Using template from %s", templateUrl)
+		if err := getter.GetAny(templateDir, templateUrl); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
```

Comment:
Can you not pass the template URL directly to `--template-url` of boilerplate? It should know how to download templates...

Reply:
I can try to pass template url but we will lose function for "detection of last release tag", not sure how important it is for template url